File: frontend/lib/intent.ts
"use client";

// ── Deterministic command parsing — GET /api/intent/parse?text=... ──────────
//
// Server-side wraps the `intent` crate (NFST/rustfst grammar match, not an
// LLM) — see crates/server/src/routes/intent.rs. A parse is a *proposal*:
// callers surface the recognized command distinctly and let the user pick
// it, then route to the same already-authorized UI flow a manual click
// would use (StatusPanel's Manage dropdown, OwnershipPanel, NeedsAction,
// the Invite modal) — never execute anything directly off free text.
//
// `target_session`/actor names (invitee, transfer recipient) come back as
// raw strings from the grammar — resolution against real sessions/actors is
// a *separate* step the server does against the caller's own membership
// (never a global lookup, same boundary as Teammates/Search), surfaced here
// as `resolution.target_session` / `resolution.actor`. A name that doesn't
// resolve isn't an error — the caller just shows the raw text.

import { authFetch } from "./auth";
import { API_BASE } from "./env";

export type Intent =
  | { kind: "pause" }
  | { kind: "resume" }
  | { kind: "archive" }
  | { kind: "approve" }
  | { kind: "deny" }
  | { kind: "claim" }
  | { kind: "invite"; role: "owner" | "collaborator" | "observer" | "agent"; invitee: string | null; ttl_secs: number | null }
  | { kind: "transfer_ownership"; to: string }
  | { kind: "navigate" }
  | { kind: "attach_agent"; name: string | null; ttl_secs: number | null };

export type NameResolution =
  | { status: "matched"; id: string; name: string; email?: string }
  | { status: "ambiguous"; candidates: { id: string; name: string; email?: string }[] }
  | { status: "not_found" };

export interface ParsedIntent {
  intent: Intent;
  target_session: string | null;
  resolution: {
    target_session?: NameResolution;
    actor?: NameResolution;
  };
}

const EMPTY: ParsedIntent | null = null;

/** Server floors identically to lib/search.ts's search() — skip the
 *  round-trip for input too short to plausibly match a grammar. */
export async function parseIntent(text: string): Promise<ParsedIntent | null> {
  if (text.trim().length < 3) return EMPTY;
  const res = await authFetch(`${API_BASE}/intent/parse?text=${encodeURIComponent(text)}`);
  if (!res.ok) return EMPTY;
  const data = await res.json();
  if (!data.intent) return EMPTY;
  return { intent: data.intent, target_session: data.target_session ?? null, resolution: data.resolution ?? {} };
}

File: crates/server/src/routes/intent.rs
//! `GET /api/intent/parse` — deterministic governance-command parsing, the
//! server-side outlet for the `intent` crate (NFST/rustfst grammar match,
//! not an LLM). Read-only and side-effect-free: it never mutates anything
//! and has no opinion on authorization — a parsed intent is a *proposal* a
//! caller (CommandPalette, the message composer) surfaces to the user, who
//! still drives the actual action through its normal, already-authorized
//! REST path. See `crates/intent/src/lib.rs`'s doc comment for the full
//! determinism/auditability rationale.
//!
//! `intent::parse_intent` is deliberately DB-agnostic — it extracts raw
//! *names* (a target session, an invitee/transfer-recipient actor) without
//! resolving them to real IDs. That resolution happens here, where DB
//! access already lives, using the exact same membership boundaries as the
//! rest of this app: target-session name matching is scoped to sessions the
//! caller belongs to (mirrors the CLI's own `resolve_session_id`), actor
//! name matching is scoped to co-members (mirrors the Teammates directory /
//! search's actor scoping — see `db::actors::list_teammates`). Resolution
//! only enriches the proposal (a name that doesn't resolve just stays an
//! unresolved string the frontend shows as-is); it never blocks a parse or
//! substitutes a guess.
//!
//! Auth-gated the same as every other `/api/*` route (`require_sp_auth`).

use std::sync::Arc;

use axum::{extract::{Query, State}, response::IntoResponse, routing::get, Json, Router};
use serde::{Deserialize, Serialize};
use sqlx::PgPool;

use intent::Intent;

use crate::state::AppState;
use autometrics::autometrics;

pub fn router() -> Router<Arc<AppState>> {
    Router::new().route("/parse", get(parse))
}

#[derive(Deserialize)]
struct ParseQuery {
    text: String,
}

#[autometrics]
async fn parse(
    headers:      axum::http::HeaderMap,
    Query(q):     Query<ParseQuery>,
    State(state): State<Arc<AppState>>,
) -> impl IntoResponse {
    let actor_id = match crate::auth::require_sp_auth(&state.db, &headers).await {
        Ok(id)   => id,
        Err(res) => return res,
    };

    let Some(parsed) = intent::parse_intent(&q.text) else {
        return Json(serde_json::json!({ "intent": null, "target_session": null, "resolution": {} })).into_response();
    };

    let mut resolution = serde_json::Map::new();
    if let Some(name) = &parsed.target_session {
        let outcome = resolve_target_session(&state.db, &actor_id, name).await;
        resolution.insert("target_session".into(), serde_json::to_value(outcome).unwrap());
    }
    let actor_name = match &parsed.intent {
        Intent::Invite { invitee: Some(name), .. } => Some(name.as_str()),
        Intent::TransferOwnership { to }            => Some(to.as_str()),
        _                                            => None,
    };
    if let Some(name) = actor_name {
        let outcome = resolve_actor(&state.db, &actor_id, name).await;
        resolution.insert("actor".into(), serde_json::to_value(outcome).unwrap());
    }

    Json(serde_json::json!({
        "intent":         parsed.intent,
        "target_session": parsed.target_session,
        "resolution":     resolution,
    })).into_response()
}

// ── Name resolution ──────────────────────────────────────────────────────

#[derive(Serialize)]
#[serde(tag = "status", rename_all = "snake_case")]
enum NameResolution {
    Matched {
        id: String,
        name: String,
        /// Only ever populated by `resolve_actor` — sessions have no email.
        /// Lets a resolved co-member's *real* email prefill the Invite
        /// modal's email field instead of just their display name.
        #[serde(skip_serializing_if = "Option::is_none")]
        email: Option<String>,
    },
    Ambiguous { candidates: Vec<NameHit> },
    NotFound,
}

#[derive(Serialize)]
struct NameHit {
    id:    String,
    name:  String,
    #[serde(skip_serializing_if = "Option::is_none")]
    email: Option<String>,
}

async fn resolve_target_session(pool: &PgPool, viewer_actor_id: &str, raw_name: &str) -> NameResolution {
    match db::sessions::list_by_actor(pool, viewer_actor_id).await {
        Ok(sessions) => resolve_by_name(raw_name, sessions.into_iter().map(|s| (s.id, s.name, None))),
        Err(_)       => NameResolution::NotFound,
    }
}

async fn resolve_actor(pool: &PgPool, viewer_actor_id: &str, raw_name: &str) -> NameResolution {
    match db::actors::list_teammates(pool, viewer_actor_id).await {
        Ok(teammates) => resolve_by_name(raw_name, teammates.into_iter().map(|t| (t.id, t.name, t.email))),
        Err(_)        => NameResolution::NotFound,
    }
}

/// Exact case-insensitive name match wins outright; otherwise fall back to
/// a substring match. Zero hits is `NotFound`, more than one at whichever
/// tier resolved is `Ambiguous` — never silently pick one.
fn resolve_by_name(raw: &str, items: impl Iterator<Item = (String, String, Option<String>)>) -> NameResolution {
    let items: Vec<(String, String, Option<String>)> = items.collect();
    let needle = raw.to_lowercase();

    let exact: Vec<&(String, String, Option<String>)> = items.iter().filter(|(_, n, _)| n.to_lowercase() == needle).collect();
    let tier = if !exact.is_empty() {
        exact
    } else {
        items.iter().filter(|(_, n, _)| n.to_lowercase().contains(&needle)).collect()
    };

    match tier.as_slice() {
        []       => NameResolution::NotFound,
        [(id, name, email)] => NameResolution::Matched { id: id.clone(), name: name.clone(), email: email.clone() },
        many     => NameResolution::Ambiguous {
            candidates: many.iter().map(|(id, name, email)| NameHit { id: id.clone(), name: name.clone(), email: email.clone() }).collect(),
        },
    }
}
